### PR TITLE
Simplify counting cards and make tests more readable.

### PR DESCRIPTION
**Pull request by TimSmole**
Mathematically counting cards can be done by subtracting 2/3 from each card, which can greatly simplify the code.
Added a few more unit tests for edge cases and made them a bit more readable.

### CODE REVIEW

File: tarok/src/cards.cpp [hunk 1]
```diff
@@ -117,24 +117,11 @@ void Shuffle(std::vector<open_spiel::Action>* actions, std::mt19937&& rng) {
...
-  int points = 0;
-  // count cards in batches of three
-  for (int i = 0; i < actions.size() / 3 * 3; i += 3) {
-    points +=
-        (deck.at(actions.at(i)).points + deck.at(actions.at(i + 1)).points +
-         deck.at(actions.at(i + 2)).points) -
-        2;
+  float points = 0;
+  for (int i = 0; i < actions.size(); i++) {
+    points += static_cast<float>(deck.at(actions.at(i)).points) - 0.666;
```
Comment (inejc) [inline tarok/src/cards.cpp]: This could simply be `points += deck.at(actions.at(i)).points - 0.666f;` methinks.

File: tarok/src/cards.cpp [hunk 2]
```diff
@@ -117,24 +117,11 @@ void Shuffle(std::vector<open_spiel::Action>* actions, std::mt19937&& rng) {
 
 int CardPoints(const std::vector<open_spiel::Action>& actions,
```
Comment (inejc) [inline tarok/src/cards.cpp:118]: Maybe we should add a short comment at the top to indicate that subtracting `2/3` from each card is equivalent to counting in threes with a 2-point deduction just for clarity?

File: tarok/test/tarok_utils.h [hunk 3]
```diff
@@ -17,4 +18,7 @@ bool AllActionsInOtherActions(
     const std::vector<open_spiel::Action>& actions,
     const std::vector<open_spiel::Action>& other_actions);
 
+open_spiel::Action ActionFromCardName(std::string short_name,
```
Comment (inejc) [inline tarok/test/tarok_utils.h]: Did you mean `long_name` instead of `short_name`?

File: tarok/test/tarok_utils.cpp [hunk 4]
```diff
@@ -28,4 +28,13 @@ bool AllActionsInOtherActions(
   return true;
 }
 
+open_spiel::Action ActionFromCardName(std::string long_name,
```
Comment (inejc) [inline tarok/test/tarok_utils.cpp]: Not really important but should probably be named with the same convention as some existing similar methods like `ActionToCard`, e.g. `CardNameToAction`.

File: tarok/test/tarok_utils.cpp [hunk 5]
```diff
@@ -28,4 +28,13 @@ bool AllActionsInOtherActions(
...
+                                      const std::array<Card, 54>& deck) {
+  for (int i = 0; i < deck.size(); i++) {
+    if (deck.at(i).long_name == long_name) {
```
Comment (inejc) [inline tarok/test/tarok_utils.cpp]: Maybe `if (deck.at(i).long_name == long_name) return i;` if it goes into a single line.

File: tarok/test/tarok_utils.cpp [hunk 6]
```diff
@@ -28,4 +28,13 @@ bool AllActionsInOtherActions(
...
+      return i;
+    }
+  }
+  throw "Invalid long_name!";
```
Comment (inejc) [inline tarok/test/tarok_utils.cpp]: Use `open_spiel::SpielFatalError`.

File: tarok/test/cards_tests.cpp [hunk 7]
```diff
@@ -53,9 +54,31 @@ TEST_F(CardsTests, TestCountCards) {
   std::iota(all_card_actions.begin(), all_card_actions.end(), 0);
   EXPECT_EQ(tarok::CardPoints(all_card_actions, deck), 70);
```
Comment (inejc) [inline tarok/test/cards_tests.cpp]: This is my mistake, may I please ask you to remove all `tarok::` here as we are in the `tarok` namespace?

File: tarok/test/cards_tests.cpp [hunk 8]
```diff
@@ -53,9 +54,31 @@ TEST_F(CardsTests, TestCountCards) {
...
   EXPECT_EQ(tarok::CardPoints({}, deck), 0);
-  EXPECT_EQ(tarok::CardPoints({13, 20, 34}, deck), 6);
-  EXPECT_EQ(tarok::CardPoints({13, 20, 34, 36}, deck), 9);
-  EXPECT_EQ(tarok::CardPoints({13, 20, 34, 36, 53}, deck), 14);
+
+  auto cards = {ActionFromCardName("XIV", deck)};
```
Comment (inejc) [inline tarok/test/cards_tests.cpp]: I think I would rename `cards` to `card_actions`.

File: tarok/test/cards_tests.cpp [hunk 9]
```diff
@@ -53,9 +54,31 @@ TEST_F(CardsTests, TestCountCards) {
...
+  cards = {ActionFromCardName("XIV", deck), ActionFromCardName("Mond", deck),
+           ActionFromCardName("Jack of Diamonds", deck)};
+  EXPECT_EQ(tarok::CardPoints(cards, deck), 6);
+
+  cards = {ActionFromCardName("XIV", deck), ActionFromCardName("Mond", deck),
+           ActionFromCardName("Jack of Diamonds", deck),
+           ActionFromCardName("Queen of Diamonds", deck)};
+  EXPECT_EQ(tarok::CardPoints(cards, deck), 9);
+
+  cards = {ActionFromCardName("XIV", deck), ActionFromCardName("Mond", deck),
```
Comment (inejc) [inline tarok/test/cards_tests.cpp]: Maybe if `ActionFromCardName` would take a vector of strings (of card names) and return a vector of actions this would make the tests even more readable. This is just a suggestion so feel free to ignore :).

**Review by inejc — COMMENTED** [review]
This is pretty nice! Some small comments :).

File: tarok/src/cards.cpp (hunk 1 above)
Comment (TimSmole) [reply tarok/src/cards.cpp]: Did some minor refactoring.

File: tarok/src/cards.cpp (hunk 2 above)
Comment (TimSmole) [reply tarok/src/cards.cpp:118]: Added a short description.

File: tarok/test/tarok_utils.h (hunk 3 above)
Comment (TimSmole) [reply tarok/test/tarok_utils.h]: Yep, my mistake. At first I used short_name but then changed it to long_name since it's more readable.

File: tarok/test/tarok_utils.cpp (hunk 4 above)
Comment (TimSmole) [reply tarok/test/tarok_utils.cpp]: Change to CardLongNameToAction.

File: tarok/test/tarok_utils.cpp (hunk 5 above)
Comment (TimSmole) [reply tarok/test/tarok_utils.cpp]: Done.

File: tarok/test/tarok_utils.cpp (hunk 6 above)
Comment (TimSmole) [reply tarok/test/tarok_utils.cpp]: Done.

File: tarok/test/cards_tests.cpp (hunk 7 above)
Comment (TimSmole) [reply tarok/test/cards_tests.cpp]: Done.

File: tarok/test/cards_tests.cpp (hunk 9 above)
Comment (TimSmole) [reply tarok/test/cards_tests.cpp]: Added CardLongNamesToActions that takes vector of strings and returns a vector of actions. Implementation is not really optimised, since it uses CardLongNameToAction which loops over the deck for each card to find its index. Nevertheless this is only used for testing so I believe this shouldn't be a problem.

File: tarok/test/tarok_utils.h [hunk 10]
```diff
@@ -17,4 +18,10 @@ bool AllActionsInOtherActions(
     const std::vector<open_spiel::Action>& actions,
     const std::vector<open_spiel::Action>& other_actions);
 
+open_spiel::Action CardLongNameToAction(std::string long_name,
```
Comment (inejc) [inline tarok/test/tarok_utils.h]: `long_name` could also be `const std:string&`.

**Review by inejc — APPROVED** [review]
